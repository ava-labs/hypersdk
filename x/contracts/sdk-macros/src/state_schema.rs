// Copyright (C) 2024, Ava Labs, Inc. All rights reserved.
// See the file LICENSE for licensing terms.

use proc_macro2::TokenStream;
use quote::quote;
use syn::{
    parse::{Parse, ParseStream},
    spanned::Spanned,
    token, Attribute, Error, Fields, Ident, Token, Type, Visibility,
};

pub fn impl_state_schema(
    key_pairs: impl IntoIterator<Item = KeyPair>,
) -> Result<TokenStream, Error> {
    key_pairs.into_iter().enumerate().try_fold(
        quote! {},
        |mut token_stream,
         (
            i,
            KeyPair {
                key_comments,
                key_vis,
                key_type_name,
                key_fields,
                value_type,
            },
        )| {
            let i = u8::try_from(i).map_err(|_| {
                Error::new(
                    key_type_name.span(),
                    "Cannot exceed `u8::MAX + 1` keys in a state-schema",
                )
            })?;

            token_stream.extend(Some(quote! {
                #(#key_comments)*
                #[derive(Copy, Clone, wasmlanche::bytemuck::NoUninit)]
                #[bytemuck(crate = "wasmlanche::bytemuck")]
                #[repr(C, packed)]
                #key_vis struct #key_type_name #key_fields;

                wasmlanche::prefixed_key_size_check!(wasmlanche::macro_types, #key_type_name);

                unsafe impl wasmlanche::macro_types::Schema for #key_type_name {
                    type Value = #value_type;

                    fn prefix() -> u8 {
                        #i
                    }
                }
            }));

            Ok(token_stream)
        },
    )
}

#[derive(Debug)]
pub struct KeyPair {
    key_comments: Vec<Attribute>,
    key_vis: Visibility,
    key_type_name: Ident,
    key_fields: Fields,
    value_type: Type,
}

impl Parse for KeyPair {
    fn parse(input: ParseStream) -> syn::Result<Self> {
        let key_comments = input.call(Attribute::parse_outer)?;
        let key_vis = input.parse::<Visibility>()?;
        let key_type_name: Ident = input.parse()?;
        let lookahead = input.lookahead1();

        let key_fields = if lookahead.peek(token::Paren) {
            let fields = input.parse()?;
            Fields::Unnamed(fields)
        } else if lookahead.peek(token::Brace) {
            Fields::Named(input.parse()?)
        } else {
            Fields::Unit
        };

        if let Fields::Named(named) = key_fields {
            return Err(Error::new(
                named.span(),
                "types with named fields are not supported",
            ));
        }

        input.parse::<Token![=>]>()?;
        let value_type = input.parse()?;

        Ok(Self {
            key_comments,
            key_vis,
            key_type_name,
            key_fields,
            value_type,
        })
    }
}
