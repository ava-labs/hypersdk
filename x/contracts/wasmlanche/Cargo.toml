[package]
name = "wasmlanche"
version = "0.1.0"
edition = "2021"

[dependencies]
borsh = { version = "1.5.1", features = ["derive"] }
bytemuck = { version = "1.17.0", features = ["derive"] }
cfg-if = "1.0.0"
displaydoc = { version = "0.2.5", default-features = false }
hashbrown = "0.14.5"
sdk-macros = { workspace = true }
simulator = { workspace = true, optional = true }
thiserror = { workspace = true }

[dev-dependencies]
env_logger = "0.11.5"

[target.'cfg(not(target_arch = "wasm32"))'.dev-dependencies]
criterion = "0.5.1"
iai = "0.1.1"
pprof = { version = "0.13.0", features = ["criterion", "flamegraph"] }
wasmlanche-test = { path = "wasmlanche-test" }

[features]
bindings = []
build = ["std"]
debug = ["std"]
std = []
test = ["std", "simulator"]
simulator = ["std", "dep:simulator"]

[[bench]]
name = "call_contract_time"
harness = false

[[bench]]
name = "call_contract_mem"
harness = false
