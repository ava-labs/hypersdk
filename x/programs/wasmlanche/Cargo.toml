[package]
name = "wasmlanche"
version = "0.1.0"
edition = "2021"

[dependencies]
borsh = { version = "1.5.1", features = ["derive"] }
bytemuck = { version = "1.17.0", features = ["derive"] }
cfg-if = "1.0.0"
displaydoc = { version = "0.2.5", default-features = false }
hashbrown = "0.14.5"
sdk-macros = { workspace = true }
simulator = { workspace = true, optional = true }
thiserror = { workspace = true }

[dev-dependencies]

[target.'cfg(not(target_arch = "wasm32"))'.dev-dependencies]
wasmtime = "14"

[features]
bindings = []
build = ["std"]
debug = ["std"]
std = []
test = ["std", "simulator"]
simulator = ["std", "dep:simulator"]
