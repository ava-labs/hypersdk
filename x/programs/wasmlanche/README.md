# Wasmlanche-sdk

This is the SDK to write web-assembly based smart-contracts for a HyperVM with smart-contracts enabled.

First and foremost, this SDK is under active development and keeping the documentation up to date with rapidly changing code can be difficult. If you see any discrepencies, please feel free to open up a GitHub issue, or even submit a PR.

### Concepts

This SDK _should_ be fairly easy to use, you just have to learn a couple of simple concepts.

Any Rust that compiles to the `wasm32-unknown-unknown` target should be valid. The SDK provides all the necessary `extern` calls into the runtime and provides safe-wrappers around these calls. It also provides a couple of macros to simplify _state_ and _function_ definitions. You can think of the `state_schema` just like you would a schema definition for a relational-database. And _public_ functions can be thought of as entrypoints with access to the particular state-space belonging to the smart-contract.

For further details, please see the generated rust-docs. You can generate them locally by running `cargo doc --open`.
