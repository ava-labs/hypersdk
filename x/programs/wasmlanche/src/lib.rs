// Copyright (C) 2024, Ava Labs, Inc. All rights reserved.
// See the file LICENSE for licensing terms.

#![deny(clippy::pedantic)]
// "build" and "debug" features enable std, so does `test`
#![cfg_attr(all(not(feature = "std"), not(test)), no_std)]

//! Welcome to the wasmlanche! This SDK provides a set of tools to help you write
//! your smart-contracts in Rust to be deployed and run on a `HyperVM`.
//!
//! There are three main concepts that you need to understand to write a smart-contract:
//! 1. **State**  
//!    State is the data that is stored on the blockchain. It also follows a schema that you specify with the [`state_schema!`] macro.
//!    <br><br>
//!
//! 2. **Public Functions**  
//!    These are the entry-points of your program. They are annotated with the [`#[public]`](crate::public) attribute.
//!    <br><br>
//!
//! 3. **Context**  
//!    The [Context] provides all access to the outer context of the execution. It is also used to access and set state with the keys defined by your schema.
//!    <br><br>
//! ## Example
//! ```
//! use wasmlanche::{public, state_schema, Address, Context};
//!
//! type Count = u64;
//!
//! state_schema! {
//!     /// Counter for each address.
//!     Counter(Address) => Count,
//! }
//!
//! /// Gets the count at the address.
//! #[public]
//! pub fn get_value(context: &mut Context, of: Address) -> Count {
//!     context
//!         .get(Counter(of))
//!         .expect("state corrupt")
//!         .unwrap_or_default()
//! }
//!
//! /// Increments the count at the address by the amount.
//! #[public]
//! pub fn inc(context: &mut Context, to: Address, amount: Count) {
//!     let counter = amount + get_value(context, to);
//!
//!     context
//!         .store_by_key(Counter(to), counter)
//!         .expect("serialization failed");
//! }
//!
//! # fn main() {}
//! ```
//!
//! ## Hint
//! Use the [dbg!] macro when testing your program, along with the `-- --nocapture` argument to your `cargo test` command.

#[cfg(feature = "build")]
pub mod build;
#[cfg(feature = "simulator")]
pub mod simulator;

mod context;
mod host;
mod memory;
mod state;
mod types;

#[cfg(feature = "debug")]
mod logging;
#[cfg(not(feature = "debug"))]
mod logging {
    pub fn log(_msg: &str) {}
    pub fn register_panic() {}
}

pub use self::{
    context::{Context, ExternalCallError, Injected},
    state::{macro_types, Error},
    types::{Address, Gas, Id, ProgramId, ID_LEN},
};
#[doc(hidden)]
pub use self::{
    logging::{log, register_panic},
    memory::HostPtr,
};

pub use sdk_macros::{public, state_schema};

// re-exports
pub use borsh;
pub use bytemuck;
