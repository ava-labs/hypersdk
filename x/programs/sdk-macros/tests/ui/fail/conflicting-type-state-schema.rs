// Copyright (C) 2024, Ava Labs, Inc. All rights reserved.
// See the file LICENSE for licensing terms.

use wasmlanche::state_schema;

struct Key;

state_schema! {
    Key => u8,
}

fn main() {}
