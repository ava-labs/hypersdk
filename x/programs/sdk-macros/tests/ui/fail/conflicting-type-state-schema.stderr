error[E0428]: the name `Key` is defined multiple times
 --> tests/ui/fail/conflicting-type-state-schema.rs:8:1
  |
6 | struct Key;
  | ----------- previous definition of the type `Key` here
7 |
8 | state_schema! {
  | ^^^^^^^^^^^^^ `Key` redefined here
  |
  = note: `Key` must be defined only once in the type namespace of this module
  = note: this error originates in the macro `state_schema` (in Nightly builds, run with -Z macro-backtrace for more info)
